force-app/main/default/classes/DMLOperations.cls

```diff
@@ -17,204 +17,337 @@
-            // Add the Opportunity to the list
+    // Delete the Leads
+    if (leads.size() > 0) {
+      delete leads;
     }
-    
-    /**
-     * Question 9
-	 * Upserts an Account with the specified name.
-	 * If an Account with the given name already exists, update the description to "Updated Account".
-	 * If no such Account exists, it creates a new one and sets the description to "New Account".
-	 * @param accountName The name of the Account to be upserted.
-	 * @return The Account that was upserted.
-	 */
-	public static Account upsertAccount(String accountName) {
-		// Query for the Account
-		List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
-
-		// if the Account exists, update the description
-
-		// otherwise, create a new Account and set the description
-
-		// Upsert the Account
-		return null;
-	}
-
-    /**
-     * Question 10
-	 * Manage the association between Contacts and Accounts
-	 * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
-	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
-     * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
-     * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
-	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
-	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
-
-    /**
-     * Question 11
-     * Insert a list of Leads and then deletes them.
-     * @param leadNames The list of Lead names to be inserted.
-     */
-    public static void insertAndDeleteLeads(List<String> leadNames) {
-        // Loop through each Lead name
-
-        // Insert the Leads
-
-        // Delete the Leads
+  }
+
+  /**
+   * Question 12
+   * Dynamically create the number of cases and relate it to the accountId, then deletes them.
+   * @param accountId The account for which the cases will be created and then deleted.
+   * @param numOfCases The number of cases to create.
+   */
+  public static void createAndDeleteCases(
+    String accountId,
+    Integer numOfCases
+  ) {
+    // Insert the cases
+
+    // Delete the cases
+
+    List<Case> casesToInsert = new List<Case>();
+    for (Integer i = 0; i < numOfCases; i++) {
+      Case newCase = new Case(
+        AccountId = accountId,
+        Subject = 'New Case ' + (i + 1),
+        Status = 'New',
+        Origin = 'Phone',
+        Priority = 'Medium'
+      );
+      casesToInsert.add(newCase);
```

Comment:
Try to use the VsCode extension Indent Rainbow to line up your brackets and spaces.
https://marketplace.visualstudio.com/items?itemName=oderwat.indent-rainbow